File: android_mobile/src/main/java/com/alexstyl/specialdates/upcoming/UpcomingEventsFreeUserAdRules.java
package com.alexstyl.specialdates.upcoming;

final class UpcomingEventsFreeUserAdRules implements UpcomingEventsAdRules {

    private static final int DAY_INTERVAL = 2;

    private int lastAdIndex = 0;
    private boolean monthHandled = false;

    @Override
    public boolean shouldAppendAdForEndOfMonth(int index) {
        return index - lastAdIndex >= DAY_INTERVAL;
    }

    @Override
    public void onNewMonthAdded(int index) {
        monthHandled = false;
        lastAdIndex = index;
    }

    @Override
    public boolean shouldAppendAdAt(int index) {
        return !monthHandled && index - lastAdIndex >= DAY_INTERVAL;
    }

    @Override
    public void onNewAdAdded(int index) {
        monthHandled = true;
        lastAdIndex = index;
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/upcoming/UpcomingEventsFreeUserAdRules.kt
package com.alexstyl.specialdates.upcoming

internal class UpcomingEventsFreeUserAdRules : UpcomingEventsAdRules {

    private var adAdded: Boolean = false

    override fun shouldAppendAd(): Boolean {
        return !this.adAdded
    }

    override fun onNewAdAdded() {
        this.adAdded = true
    }
}
